### PR TITLE
Create view of edit client application type

### PR DESCRIPTION
**Pull request by nehajain008**
Story AB#15298
Task AB#15969

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs [hunk 1]
```diff
@@ -225,5 +225,41 @@ public IActionResult HostingTypeData()
...
+
+            model = new ClientApplicationTypeSectionModel(catalogueItem);
+
+            if (!Convert.ToBoolean(model.AddedClientApplicationType))
+            {
+                ModelState.AddModelError(" ", "Add an application type");
+                var solution = await solutionsService.GetSolution(solutionId);
+                model = new ClientApplicationTypeSectionModel(solution);
+                return View(model);
+            }
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs:253]: This action method has the same issues as those described in #278, such as:

-  the `model` that is received as a parameter is ignored
- the validation should probably be moved to a validator

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System.Threading.Tasks;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Catalogue.Models;
+using NHSD.GPIT.BuyingCatalogue.Framework.Extensions;
+using NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Marketing.Models;
+using NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Marketing.Models.ClientApplicationType;
+using NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Solutions.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models
+{
+    public class ClientApplicationTypeSectionModel : MarketingBaseModel
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:13]: `sealed`

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        public ClientApplicationTypeSectionModel(CatalogueItem catalogueItem)
+            : base(catalogueItem)
+        {
+            if (catalogueItem is null)
+                throw new ArgumentNullException(nameof(catalogueItem));
+
+            SolutionName = catalogueItem?.Name;
+            IsBrowserBased = Convert.ToBoolean(CatalogueItem.Solution?.GetClientApplication()?.ClientApplicationTypes.Any(x => x.EqualsIgnoreCase("browser-based")));
+            IsNativeMobile = Convert.ToBoolean(CatalogueItem.Solution?.GetClientApplication()?.ClientApplicationTypes.Any(x => x.EqualsIgnoreCase("native-mobile")));
+            IsNativeDesktop = Convert.ToBoolean(CatalogueItem.Solution?.GetClientApplication()?.ClientApplicationTypes.Any(x => x.EqualsIgnoreCase("native-desktop")));
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:29]: Can we use something other than `x` for the lambda parameter?

There's no need to use `Convert.ToBoolean` here either, instead we could use `expression ?? false`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        public string SolutionName { get; set; }
+
+        public virtual bool IsBrowserBased { get; set; }
+
+        public virtual bool IsNativeMobile { get; set; }
+
+        public virtual bool IsNativeDesktop { get; set; }
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:38]: Don't need to be `virtual`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        public override bool? IsComplete => Convert.ToBoolean(IsBrowserBased) || Convert.ToBoolean(IsNativeMobile) || Convert.ToBoolean(IsNativeDesktop);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:40]: No need for `Convert.ToBoolean`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        public bool? AddedClientApplicationType => IsComplete == true ? true : null;
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:42]: Why nullable? Can we not just return `false` is not complete? In which case, we can just do `IsComplete.GetValueOrDefault()`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        public FeatureCompletionStatus StatusClientApplicationType() =>
+           Convert.ToBoolean(IsComplete)
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:45]: IsComplete.GetValueOrDefault()

**Review by bjss-chris — CHANGES_REQUESTED** [review]
The follows a similar pattern to #278 in places and as a result, it has some of the same issues. I pushed a couple of commits with suggested changes to @RahulJha-mastek's branch on #278 so I'd recommend having a look at my suggested changes for #278 – you can probably apply similar fixes here.

Let me know if you have any questions.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs (hunk 1 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs]: Changes done as described in #278

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 2 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs:13]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 3 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs]: Done.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 4 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs]: Removed these properties.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 5 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs]: Done.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 6 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs]: Done.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs (hunk 7 above)
Comment (nehajain008) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/ClientApplicationTypeSectionModel.cs]: Done.

**Comment by nehajain008** [comment]
> 
> 
> The follows a similar pattern to #278 in places and as a result, it has some of the same issues. I pushed a couple of commits with suggested changes to @RahulJha-mastek's branch on #278 so I'd recommend having a look at my suggested changes for #278 – you can probably apply similar fixes here.
> 
> Let me know if you have any questions.

Done.

**Comment by bjss-chris** [comment]
Fixed up a couple of issues with the existing functionality and content.

There's some tech debt here – added some TODO comments to review in the near future.

**Pull request merged**